File: 02.lua
local target = {red = 12, green = 13, blue = 14}
local sum, pwr = 0, 0

for id, games in io.open("02.txt"):read("*a"):gmatch("Game (%d+): (.-)\n") do
    local g, cur = true, {}

    for game in games:gmatch("[^;]+") do
        for n, color in game:gmatch("(%d+) (%w+)") do
            n = tonumber(n)
            cur[color] = math.max(cur[color] or 0, n)
            g = g and n <= target[color]
        end
    end

    sum = sum + (g and tonumber(id) or 0)
    pwr = pwr + cur.red * cur.green * cur.blue
end

print(sum, pwr)

File: 02.py
#!/usr/bin/env python3

from aochelper import *

filename = sys.argv[1] if len(sys.argv) > 1 else '02.txt'
inp = open(filename).read().splitlines()

def solve(inp):
    target = {'red': 12, 'green': 13, 'blue': 14}
    p1 = p2 = 0
    for line in inp:
        vals = {'red': 0, 'green': 0, 'blue': 0}
        id = line.split(': ')[0].split(' ')[-1]
        subsets = line.split(': ')[1].split('; ')
        game = [subset.split(', ') for subset in subsets]
        ok = True
        for subset in game:
            for cube in subset:
                n, color = cube.split()
                n = int(n)
                vals[color] = max(vals[color], n)
                if n > target[color]:
                    ok = False
        if ok:
            p1 += int(id)
        p2 += math.prod(vals.values())

    return p1, p2

print(*solve(inp))
